api4/channel.go

```diff
@@ -1112,32 +1113,46 @@ func addChannelMember(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// Bypass permissions check if you are already joined to the channel
+	if channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP {
+		c.Err = model.NewAppError("addUserToChannel", "api.channel.add_user_to_channel.type.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+
+	isNewMembership := false
 	if _, err = c.App.GetChannelMember(member.ChannelId, member.UserId); err != nil {
-		// Check join permission if adding yourself, otherwise check manage permission
-		if channel.Type == model.CHANNEL_OPEN {
-			if member.UserId == c.App.Session.UserId {
-				if !c.App.SessionHasPermissionToTeam(c.App.Session, channel.TeamId, model.PERMISSION_JOIN_PUBLIC_CHANNELS) {
-					c.SetPermissionError(model.PERMISSION_JOIN_PUBLIC_CHANNELS)
-					return
-				}
-			} else {
-				if !c.App.SessionHasPermissionToChannel(c.App.Session, channel.Id, model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS) {
-					c.SetPermissionError(model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS)
-					return
-				}
-			}
+		if err.Id == store.MISSING_CHANNEL_MEMBER_ERROR {
+			isNewMembership = true
+		} else {
+			c.Err = err
+			return
 		}
+	}
 
-		if channel.Type == model.CHANNEL_PRIVATE && !c.App.SessionHasPermissionToChannel(c.App.Session, channel.Id, model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS) {
-			c.SetPermissionError(model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS)
-			return
+	isSelfAdd := member.UserId == c.App.Session.UserId
+
+	if channel.Type == model.CHANNEL_OPEN {
+		if !isSelfAdd {
+			if !c.App.SessionHasPermissionToChannel(c.App.Session, channel.Id, model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS) {
+				c.SetPermissionError(model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS)
+				return
+			}
+		} else if isNewMembership {
+			if !c.App.SessionHasPermissionToTeam(c.App.Session, channel.TeamId, model.PERMISSION_JOIN_PUBLIC_CHANNELS) {
+				c.SetPermissionError(model.PERMISSION_JOIN_PUBLIC_CHANNELS)
+				return
+			}
 		}
+		// If it is a self add and I'm already in the channel, do not check any permission
 	}
 
-	if channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP {
-		c.Err = model.NewAppError("addUserToChannel", "api.channel.add_user_to_channel.type.app_error", nil, "", http.StatusBadRequest)
-		return
+	if channel.Type == model.CHANNEL_PRIVATE {
+		if !isSelfAdd || isNewMembership {
```

Comment:
Same comment as above, but suggesting:
```go
if isSelfAdd && isNewMembership {
    // check model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS
} else if isSelfAdd && !isNewMembership {
    // nothing to do, since already in the channel
} else if !isSelfAdd {
   // check model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS
}
```

The old code was correct, but the proposal above creates a parallelism between the two permission checking branches.